File: WorkFlows/Chapter07/VBCommunicationSequentialConsoleApplication/ReviewService.vb
<ExternalDataExchange()> _
Public Interface IReview
    Function CreateReview(ByVal Reviewer As String, ByVal Reviewee As String) As Boolean
    Event ReviewApproved As EventHandler(Of ExternalDataEventArgs)
    Event ReviewNotApproved As EventHandler(Of ExternalDataEventArgs)
End Interface
Public Class ReviewService : Implements IReview
    Private StrReviewer As String
    Private StrReviewee As String

    Public Function CreateReview(ByVal Reviewer As String, ByVal Reviewee As String) As Boolean Implements IReview.CreateReview
        StrReviewer = Reviewer
        StrReviewee = Reviewee
        MsgBox("Reviewer: " & StrReviewer)

        ThreadPool.QueueUserWorkItem(New System.Threading.WaitCallback(AddressOf AskForApproval))

        Return True
    End Function

    Public Event ReviewApproved(ByVal sender As Object, ByVal e As ExternalDataEventArgs) Implements IReview.ReviewApproved

    Public Event ReviewNotApproved(ByVal sender As Object, ByVal e As ExternalDataEventArgs) Implements IReview.ReviewNotApproved
    Private Sub AskForApproval(ByVal o As Object)
        If MsgBox("Do you approve the review for: " & StrReviewee & " ?", MsgBoxStyle.YesNo, "Approve review?") = MsgBoxResult.Yes Then
            RaiseEvent ReviewApproved(Nothing, Nothing)
        Else
            RaiseEvent ReviewNotApproved(Nothing, Nothing)
        End If
    End Sub
    Private Sub ApproveReview(ByVal sender As Object, ByVal e As ExternalDataEventArgs) Handles Me.ReviewApproved
        MsgBox("Reviewer: " & StrReviewer & " has approved the review for " & StrReviewee)
    End Sub
    Private Sub DoNotApproveReview(ByVal sender As Object, ByVal e As ExternalDataEventArgs) Handles Me.ReviewNotApproved
        MsgBox("Reviewer: " & StrReviewer & " has not approved the review for " & StrReviewee)
    End Sub

End Class

File: WorkFlows/Chapter07/CCommunicationSequentialConsoleApplication/ReviewService.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Workflow.Activities;
using System.Workflow.ComponentModel;
using System.Workflow.Runtime;
using System.Windows.Forms;
using System.Threading;

namespace CCommunicationSequentialConsoleApplication
{
    [Serializable]
    internal class ReviewEventArgs : ExternalDataEventArgs
    {
        private string alias;

        public ReviewEventArgs(Guid InstanceID, string alias)
            : base(InstanceID)
        {
            this.alias = alias;
        }

        public string Alias
        {
            get { return this.alias; }
            set { this.alias = value; }
        }
    }
   
    [ExternalDataExchange]
    interface IReview
    {
        Boolean CreateReview(String Reviewer, String Reviewee);
        event EventHandler<ExternalDataEventArgs  > ReviewApproved;
        event EventHandler<ExternalDataEventArgs  > ReviewNotApproved;
    }

	class ReviewService:IReview
	{
        static String StrReviewer;
        static String StrReviewee;
        public event EventHandler<ExternalDataEventArgs > ReviewApproved;
        public event EventHandler<ExternalDataEventArgs > ReviewNotApproved;

        public Boolean CreateReview(String Reviewer, String Reviewee)
        {
            StrReviewer = Reviewer;
            StrReviewee = Reviewee;
            MessageBox.Show("Reviewer: " + Reviewer);
            ThreadPool.QueueUserWorkItem (AskForApproval ,new ReviewEventArgs(WorkflowEnvironment.WorkflowInstanceId,Reviewer ));
            return true;
        }
        public void AskForApproval(Object O)
        {
            DialogResult Result;
            ReviewEventArgs  revieweargs = O as ReviewEventArgs ;
            Guid instanceId = revieweargs.InstanceId;
            string alias = revieweargs.Alias;
            Result = MessageBox.Show("Do you approve the review for " + StrReviewee + " ?", "Approval", MessageBoxButtons.YesNo);
            if (Result == DialogResult.Yes)
            {
              ReviewApproved(null, revieweargs);
            }
            else
            {
                ReviewNotApproved(null, revieweargs);
            }

        }

        
	}
}
